### ISSUE TITLE
URL state encoding

### ISSUE THREAD

**Issue by NealHumphrey** [codefordc/housing-insights#387]
After selecting several filter parameters, I want to be able to bookmark or share my findings by sending a URL to someone. When changing views in the single page app (such as changing from map view to look at a specific building) I also want my back button to behave as expected. Both of these things necessitate encoding the state of the app in the URL. 

Core functionality:
- [ ] Every time the user modifies a 'saveable state element', modify the browser url with `history.replaceState`
    - [ ] Change a filter
    - [ ] Switch sidebar type
    - [ ] Expand a sidebar
    - [ ] Select a specific building for preview mode
- [ ] Every time the user modifies a 'browsable state element', modify the browser url AND back-button history with `history.pushState`
    - [ ] Switch to building view
    - [ ] Switch to map view

# Proposed Implementation
## Status quo:
Currently our the 'state' variable (accessed via `getState()`) holds all the necessary information about the app state to recreate a specific view. However, it also holds a) other unnecessary stuff, such as 'dataLoaded' events, b) large data sets that are based on other state elements, like the filteredData object, and c) the previous state of each component in a 2 element long array. As such, we'll need to separate the state elements that are necessary for the URL from those that aren't. 

Once we have the relevant values, we need to be able to translate back and forth between the URL and the state itself. 
- First, we'll need to turn the current urlState into a url. To start with, the [jQuery.param()](http://api.jquery.com/jQuery.param/) function looks like the easiest way to do this. However, it will be quite possible that users could occasionally end up with very long URLs - potentially exceeding the [suggested limit of 2000 characters](https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers). IE is the real limiting factor here, for which we already have spotty support but need to not ignore completely. We can manage this a little bit through careful attention to the length of both variables and values, such as mapping 'Subsidy Program' names to a key index. Longer term we could store states in the database if complex urls are common. 
- Then we'll need to turn a URL into a restored state. We'll want to be careful that if lots of similar events are triggered (such as each of the selected filters) they don't re-run the relevant code multiple times which can cause performance issues - this is the same problem that is currently occurring with the 'clear filters' button that triggers the recalculation of the filtered buildings multiple times. We also want to make sure the application of the setState() for these url params occurs at the right time - for instance, we can't apply filters if dataLoaded.filterData hasn't triggered yet. 

## Proposed architecture:
Keep the current 'state' variable as is; allow a subset of url-relevant state params to be read out of the state variable when needed, and on page load parse the url and set the state based on these parameters thus triggering the relevant view-changing code. 

### Triggering filters only once:
- [ ] Add a 'setMultipleStates' function that accepts a list of key/value state pairs (`[{key: 'sidebar', "value":"buildings"},{"key":"filterValues.proj_units_tot","value":[0,100]}]`). function should iterate over every list element and assign it to the state variable **first**, and then iterate over every element a second time and publish the state via `PubSub.publish()`. However, any time `filterValues.whatever` is included it should not be published and instead a single publish should occur with `PubSub.publish("updateFilters",true)`. 
- [ ] Add an optional 'forcePublish' parameter to the setState() function - default should be false, if true than it overrides the ` else if ( state[key][0] !== value )` conditional and publishes the event via PubSub. 
- [ ] Create a new subscription to the `filterValues` event that simply creates an `"updateFilters",true` event. Use the new setState param to force publish this. 
- [ ] Change the filterUtil.publishFilteredData to instead subscribe to the updateFilters event. 

### Encoding the URL
- [ ] Create a function in housing-insights.js that extracts just the URL-relevant components from the state module. This should only include the most recent state (not the history). 
- [ ] Function that uses history.pushState and replaceState to keep the url and browser history updated using `$.param()` to encode the urlState object. 
- [ ] Subscribe the url function to all relevant state changes (try to keep this DRY with some shared config to the url-extractor function if you can)

### Regenerating the state
- [ ] function that calls setMultipleStates with the urlState object parsed from the url, which is triggered during page load but late enough in the process that the relevant items can trigger.

**Comment by NealHumphrey** [comment]
Url state encoding completed in #422. There's a couple minor enhancements to it (sidebar state, etc.) - will open a new issue for those instead.
